### PR TITLE
Bug 580015: add support for multiple bin parsers

### PR DESCRIPTION
**Pull request by betamaxbandit**
Added ability to return a list of binary parser IDs, rather than a single ID. This supports build configurations that have multiple binaries with for example cross toolchains.

Change-Id: I1b7e47bf6a86bbd9f1c6b9646d008bac9479417d

### CODE REVIEW

File: core/org.eclipse.cdt.core/.settings/.api_filters [hunk 1]
```diff
@@ -166,6 +166,22 @@
             </message_arguments>
         </filter>
     </resource>
+    <resource path="src/org/eclipse/cdt/core/build/ICBuildConfiguration.java" type="org.eclipse.cdt.core.build.ICBuildConfiguration">
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core/.settings/.api_filters:169]: With CDT 11 we can just make this a new major version of the bundle and not have to have this API filter.

WDYT?

File: core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java [hunk 2]
```diff
@@ -79,9 +80,23 @@ default String getLaunchMode() {
...
 
+	/**
+	 * Ids for the Binary Parsers to use when checking whether a file is a
+	 * binary that can be launched.
+	 *
+	 * @return binary parser ids
+	 * @throws CoreException
+	 * @since 7.5
+	 */
+	default List<String> getBinaryParserIds() throws CoreException {
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java:96]: With a new major revision (CDT 11) this method shouldn't have a default anymore.

**Review by jonahgraham — CHANGES_REQUESTED** [review]
This is migrated from https://git.eclipse.org/r/c/cdt/org.eclipse.cdt/+/194075 - so the comments from it are migrated to it.

**Comment by betamaxbandit** [comment]
Hi @jonahgraham ,
Can you review my changes to the actual code please? Note I'm still working on the tests (TestICBuildConfiguration), but ran out of time today.

File: core/org.eclipse.cdt.core.tests/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -45,7 +45,8 @@ Require-Bundle: org.eclipse.core.resources,
  org.eclipse.ltk.core.refactoring;bundle-version="3.4.0",
  org.hamcrest.core,
  org.hamcrest.library,
- com.google.gson;bundle-version="[2.8.6,3.0.0)"
+ com.google.gson;bundle-version="[2.8.6,3.0.0)",
+ org.eclipse.cdt.build.gcc.core
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core.tests/META-INF/MANIFEST.MF:49]: I think this dependency is a problem - not sure core.tests should depend on something not in cdt core.

File: core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java [hunk 4]
```diff
@@ -79,9 +80,21 @@ default String getLaunchMode() {
 	 *
 	 * @return binary parser ids
 	 * @throws CoreException
+	 * @deprecated As of 8.0 replaced by {@link ICBuildConfiguration#getBinaryParserIds}
 	 */
+	@Deprecated(since = "8.0")
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java:85]: Delete this, don't just deprecate it.

File: core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/IToolChain.java [hunk 5]
```diff
@@ -165,9 +165,20 @@ default String getTypeId() {
 	 * the toolchain.
 	 *
 	 * @return binary parser IDs for this toolchain
+	 * @deprecated As of 8.0 replaced by {@link IToolChain#getBinaryParserIds}
 	 */
+	@Deprecated(since = "8.0")
```
Comment (jonahgraham) [inline core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/IToolChain.java:170]: Delete this, don't just deprecate it.

**Review by jonahgraham — CHANGES_REQUESTED** [review]
I didn't look at tests, other than the location of them. Let me know if you need a suggestion of where they should go. It may also be appropriate to leave them there if there isn't anywhere better already.

File: core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java (hunk 4 above)
Comment (jonahgraham) [reply core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java:85]: The reason to bump the major version for this is precisely so we can delete the API without deprecation.
https://github.com/eclipse-cdt/cdt/blob/main/doc/org.eclipse.cdt.doc.isv/guide/deprecated_API_removals.html needs to be updated to list the deprecation + mitigation.

Comment (jonahgraham) [reply core/org.eclipse.cdt.core/src/org/eclipse/cdt/core/build/ICBuildConfiguration.java:85]: (should have read ) needs to be updated to list the **removal** + mitigation.

**Comment by jonahgraham** [comment]
@betamaxbandit Will you have time to look at this this week? It would be ideal to get this done in time for M3 which is 14 Nov.

**Comment by betamaxbandit** [comment]
> @betamaxbandit Will you have time to look at this this week? It would be ideal to get this done in time for M3 which is 14 Nov.

Hi Jonah,

Yes, I'll try. Thanks for the reminder.

**Comment by jonahgraham** [comment]
There is a compilation failure now BTW:

```sh
Error:  Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.5:compile (default-compile) on project org.eclipse.cdt.docker.launcher: Compilation failure: Compilation failure: 
Error:  /home/runner/work/cdt/cdt/launch/org.eclipse.cdt.docker.launcher/src/org/eclipse/cdt/internal/docker/launcher/ui/launchbar/ContainerGCCToolChain.java:[70] 
Error:  	public class ContainerGCCToolChain extends PlatformObject implements IToolChain, IToolChain2 {
Error:  	             ^^^^^^^^^^^^^^^^^^^^^
Error:  The type ContainerGCCToolChain must implement the inherited abstract method IToolChain.getBinaryParserIds()
Error: [ERROR] 1 problem (1 error)
Error:  -> [Help 1]
```

**Comment by betamaxbandit** [comment]
Hi @jonahgraham ,
I didn't mean to do that. Still getting used to the github i/f. I was trying to rebase on main, but not sure I did it correctly.

**Comment by betamaxbandit** [comment]
> Don't worry about the code cleanliness checks fails - there is a proposal to automate this in #166

Hi Jonah, ok, I won't.
I fixed the remaining problem which was the junit tests were failing. All tests pass now :-)

**Review by jonahgraham — CHANGES_REQUESTED** [review]
This looks fine, just needs a N&N entry about the improved API.

In https://github.com/eclipse-cdt/cdt/blob/main/NewAndNoteworthy/CDT-11.0.md below `Binary Parser code uses AutoCloseable` section

```
## Core Build tool chains can now return multiple binary parsers

The `ICBuildConfiguration` and `IToolChain` interfaces now have a method, `getBinaryParserIds` that allows a build configuration or tool chain to return multiple binary parsers.

See https://github.com/eclipse-cdt/cdt/pull/75

```

**Review by jonahgraham — CHANGES_REQUESTED** [review]
LGTM - needs the code cleanliness sorted out. I would like to use this PR to test #166 if I can get that done.

**Comment by betamaxbandit** [comment]
> LGTM - needs the code cleanliness sorted out. I would like to use this PR to test #166 if I can get that done.

Oh ok. I don't really understand the actual cleanliness error. It says the following files are missing, but those file don't exist for me locally - only org.eclipse.jdt.core.prefs. Am I expected to generate them?

	build/org.eclipse.cdt.build.gcc.core.tests/.settings/org.eclipse.core.resources.prefs
	build/org.eclipse.cdt.build.gcc.core.tests/.settings/org.eclipse.jdt.launching.prefs
	build/org.eclipse.cdt.build.gcc.core.tests/.settings/org.eclipse.jdt.ui.prefs
	build/org.eclipse.cdt.build.gcc.core.tests/.settings/org.eclipse.pde.api.tools.prefs
	build/org.eclipse.cdt.build.gcc.core.tests/.settings/org.eclipse.pde.prefs

Do you want me to do anything for https://github.com/eclipse-cdt/cdt/issues/166 ?

**Comment by jonahgraham** [comment]
> Do you want me to do anything for #166 ?

No.

If I can get #166 to work it will magically and automatically fix this PR. If not I will fix it up manually and merge it in a few days (ahead of RC1 on Monday)

**Review by jonahgraham — APPROVED** [review]

**Comment by betamaxbandit** [comment]
Thanks for merging @jonahgraham

**Pull request merged**